src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,237 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Christopher Hoskin
+-/
+import algebra.group.pi
+import algebra.group_power.lemmas
+
+/-!
+# Centroid homomorphisms
+
+This file defines centroid homomorphisms.
+
+We use the `fun_like` design, so each type of morphisms has a companion typeclass which is meant to
+be satisfied by itself and all stricter types.
+
+## Types of morphisms
+
+* `centroid_hom`: Maps which preserve left and right multiplication.
+
+## Typeclasses
+
+* `centroid_hom_class`
+
+## Tags
+
+centroid
+-/
+
+open function
+
+variables {F α : Type*}
+
+/-- The type of centroid homomorphisms from `α` to `α`. -/
+structure centroid_hom (α : Type*) [non_unital_non_assoc_semiring α] extends α →+ α :=
+(map_mul_left' (a b : α) : to_fun (a * b) = a * to_fun b)
+(map_mul_right' (a b : α) : to_fun (a * b) = to_fun a * b)
+
+/-- `centroid_hom_class F α` states that `F` is a type of centroid homomorphisms.
+
+You should extend this class when you extend `centroid_hom`. -/
+class centroid_hom_class (F : Type*) (α : out_param $ Type*) [non_unital_non_assoc_semiring α]
+  extends add_monoid_hom_class F α α :=
+(map_mul_left (f : F) (a b : α) : f (a * b) = a * f b)
+(map_mul_right (f : F) (a b : α) : f (a * b) = f a * b)
+
+export centroid_hom_class (map_mul_left map_mul_right)
+
+instance [non_unital_non_assoc_semiring α] [centroid_hom_class F α] :
+  has_coe_t F (centroid_hom α) :=
+⟨λ f, { to_fun := f, map_mul_left' := map_mul_left f, map_mul_right' := map_mul_right f,
+  ..(f : α →+ α) }⟩
+
+/-! ### Centroid homomorphisms -/
+
+namespace centroid_hom
+
+section non_unital_non_assoc_semiring
+
+variables [non_unital_non_assoc_semiring α]
+
+instance : centroid_hom_class (centroid_hom α) α :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by { obtain ⟨⟨_, _⟩, _⟩ := f, obtain ⟨⟨_, _⟩, _⟩ := g, congr' },
+  map_zero := λ f, f.map_zero',
+  map_add := λ f, f.map_add',
+  map_mul_left := λ f, f.map_mul_left',
+  map_mul_right := λ f, f.map_mul_right' }
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`
+directly. -/
+instance : has_coe_to_fun (centroid_hom α) (λ _, α → α) := fun_like.has_coe_to_fun
+
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf] lemma to_fun_eq_coe {f : centroid_hom α} : f.to_fun = (f : α → α) := rfl
+
+@[ext] lemma ext {f g : centroid_hom α} (h : ∀ a, f a = g a) : f = g := fun_like.ext f g h
+
+@[simp] lemma coe_to_add_monoid_hom (f : centroid_hom α) : ⇑(f : α →+ α) = f := rfl
+@[simp] lemma to_add_monoid_hom_eq_coe (f : centroid_hom α) : f.to_add_monoid_hom = f :=
+fun_like.coe_injective rfl
```

Comment:
@Vierkantor, this is no good... The rebundling that the coercion provokes means this lemma is not refl.

Reply:
I think* this will become defeq by eta for structures in Lean 4. So I'm not so worried about this if we are careful to use the coercion rather than the `.to_add_monoid_hom` projection.

\* Haven't tested it, but I know structure eta allows `equiv.symm.symm e = e := rfl`.

Reply:
Why don't we just make `centroid_hom` an old structure, which makes this problem go away?

Reply:
With the `old_structure_cmd` suggestion above, this becomes:
```suggestion
@[simp] lemma to_add_monoid_hom_eq_coe (f : centroid_hom α) : f.to_add_monoid_hom = f := rfl
```

Reply:
Should we recommend that _all_ fun_like subclasses are implemented with `old_structure_cmd`?

Reply:
The issue here is more specifically that the generic coercion is not defeq to the specific one. So only those subclasses for which the parent class has a coercion. Apart from `[add_]monoid_hom` I believe the other example is `ring_hom`.

Reply:
... or `complete_lattice_hom`: #12855

Reply:
It seems we have agreed `old_structure_cmd` is the way to go. Could you apply the `old_structure_cmd` suggestion?

Reply:
I still believe that new structures are the way to go for hom classes, so I will let Christopher decide.

Reply:
Can you give a justification for that belief?

Reply:
This PR seems to have take on a life of its own since I first submitted it. I don't have enough experience of building a library yet to have an opinion on what the right way to do things is - I just try to find something similar that already exists and imitate that.

Reply:
If there isn't agreement on the correct approach, perhaps we should accept it as is for now? Someone can then always submit a follow up PR if it causes problems later?